### PR TITLE
Additional Features

### PR DESCRIPTION
**Pull request by roblandry**
Opening a single PR for additional features. Closing others.
Uses f-strings from python 3.6.

Supported SOAP Functions
```
# ---------------------
# SERVICE_DEVICE_CONFIG
# ---------------------
LOGIN = 'SOAPLogin'
# LOGOUT = 'SOAPLogout'
REBOOT = 'Reboot'
CHECK_NEW_FIRMWARE = 'CheckNewFirmware'
# UPDATE_NEW_FIRMWARE = 'UpdateNewFirmware'
CHECK_APP_NEW_FIRMWARE = 'CheckAppNewFirmware'  # ***NEW***
CONFIGURATION_STARTED = 'ConfigurationStarted'
CONFIGURATION_FINISHED = 'ConfigurationFinished'
GET_DEVICE_CONFIG_INFO = 'GetInfo'  # ***NEW***

# BLOCK/ALLOW DEVICE
SET_BLOCK_DEVICE_ENABLE = 'SetBlockDeviceEnable'
GET_BLOCK_DEVICE_ENABLE_STATUS = 'GetBlockDeviceEnableStatus'
# ENABLE_BLOCK_DEVICE_FOR_ALL = 'EnableBlockDeviceForAll'  # deprecated?
SET_BLOCK_DEVICE_BY_MAC = 'SetBlockDeviceByMAC'

# TRAFFIC METER
GET_TRAFFIC_METER_STATISTICS = 'GetTrafficMeterStatistics'
ENABLE_TRAFFIC_METER = 'EnableTrafficMeter'
GET_TRAFFIC_METER_ENABLED = 'GetTrafficMeterEnabled'
# SET_TRAFFIC_METER_OPTIONS = 'SetTrafficMeterOptions'
GET_TRAFFIC_METER_OPTIONS = 'GetTrafficMeterOptions'

# ---------------------
# SERVICE_LAN_CONFIG_SECURITY
# ---------------------
GET_LAN_CONFIG_SEC_INFO = 'GetInfo'	 # ***NEW***

# ---------------------
# SERVICE_WAN_IP_CONNECTION
# ---------------------
GET_WAN_IP_CON_INFO = 'GetInfo'  # ***NEW***

# ---------------------
# SERVICE_PARENTAL_CONTROL
# ---------------------
LOGIN_OLD = 'Authenticate'
ENABLE_PARENTAL_CONTROL = 'EnableParentalControl'
GET_PARENTAL_CONTROL_ENABLE_STATUS = 'GetEnableStatus'
GET_ALL_MAC_ADDRESSES = 'GetAllMACAddresses'
# SET_DNS_MASQ_DEVICE_ID = 'SetDNSMasqDeviceID'
GET_DNS_MASQ_DEVICE_ID = 'GetDNSMasqDeviceID'
# DELETE_MAC_ADDRESS = 'DeleteMACAddress'

# ---------------------
# SERVICE_DEVICE_INFO
# ---------------------
GET_INFO = 'GetInfo'
GET_SUPPORT_FEATURE_LIST_XML = 'GetSupportFeatureListXML'
GET_ATTACHED_DEVICES = 'GetAttachDevice'
GET_ATTACHED_DEVICES_2 = 'GetAttachDevice2'
# SET_DEVICE_NAME_ICON_BY_MAC = 'SetDeviceNameIconByMAC'
# SET_DEVICE_NAME = 'SetNetgearDeviceName'	 # ***NEW***

# ---------------------
# SERVICE_ADVANCED_QOS
# ---------------------
SET_SPEED_TEST_START = 'SetOOKLASpeedTestStart'
GET_SPEED_TEST_RESULT = 'GetOOKLASpeedTestResult'
SET_QOS_ENABLE_STATUS = 'SetQoSEnableStatus'
GET_QOS_ENABLE_STATUS = 'GetQoSEnableStatus'
# SET_BANDWIDTH_CONTROL_OPTIONS = 'SetBandwidthControlOptions'
GET_BANDWIDTH_CONTROL_OPTIONS = 'GetBandwidthControlOptions'
GET_CURRENT_APP_BANDWIDTH = 'GetCurrentAppBandwidth'  # Not Working
GET_CURRENT_DEVICE_BANDWIDTH = 'GetCurrentDeviceBandwidth'  # Not Working
GET_CURRENT_APP_BANDWIDTH_BY_MAC = 'GetCurrentAppBandwidthByMAC'  # Not Working

# ---------------------
# SERVICE_WLAN_CONFIGURATION
# ---------------------
SET_GUEST_ACCESS_ENABLED = 'SetGuestAccessEnabled'  # 2.4G-1 R7800
GET_GUEST_ACCESS_ENABLED = 'GetGuestAccessEnabled'  # 2.4G-1 R7800/R8000
SET_GUEST_ACCESS_ENABLED_2 = 'SetGuestAccessEnabled2'  # 2.4G-1 R8000
GET_GUEST_ACCESS_ENABLED_2 = 'GetGuestAccessEnabled2'  # 2.4G-1 R8000
SET_5G_GUEST_ACCESS_ENABLED = 'Set5GGuestAccessEnabled'  # 5G-1 R7800
GET_5G1_GUEST_ACCESS_ENABLED = 'Get5GGuestAccessEnabled'  # 5G-1 R7800
GET_5G1_GUEST_ACCESS_ENABLED_2 = 'Get5G1GuestAccessEnabled'  # 5G-1 R8000
SET_5G1_GUEST_ACCESS_ENABLED_2 = 'Set5G1GuestAccessEnabled2'  # 5G-1 R8000
SET_5G_GUEST_ACCESS_ENABLED_2 = 'Set5GGuestAccessEnabled2'  # 5G-2 R8000
GET_5G_GUEST_ACCESS_ENABLED_2 = 'Get5GGuestAccessEnabled2'  # 5G-2 R8000
GET_WPA_SECURITY_KEYS = 'GetWPASecurityKeys'
GET_5G_WPA_SECURITY_KEYS = 'Get5GWPASecurityKeys'
GET_2G_INFO = 'GetInfo'
GET_5G_INFO = 'Get5GInfo'
# SET_5G_WLAN_WPA_PSK_BY_PASSPHRASE = 'Set5GWLANWPAPSKByPassphrase'
GET_AVAILABLE_CHANNEL = 'GetAvailableChannel'
# SET_GUEST_ACCESS_NETWORK = 'SetGuestAccessNetwork'
GET_GUEST_ACCESS_NETWORK_INFO = 'GetGuestAccessNetworkInfo'
# SET_5G_GUEST_ACCESS_NETWORK = 'Set5GGuestAccessNetwork'
GET_5G_GUEST_ACCESS_NETWORK_INFO = 'Get5GGuestAccessNetworkInfo'
GET_SMART_CONNECT_ENABLED = 'IsSmartConnectEnabled'	 # ***NEW***
SET_SMART_CONNECT_ENABLED = 'SetSmartConnectEnable'	 # ***NEW***

# ---------------------
# SERVICE_USER_OPTIONS_TC
# ---------------------
#SET_USER_OPTION_TC = 'SetUserOptionsTC'	 # ***NEW***
```

### CODE REVIEW

File: pynetgear/__init__.py [hunk 1]
```diff
@@ -209,309 +761,500 @@ def get_attached_devices_2(self):
...
+            "method": c.SET_SMART_CONNECT_ENABLED,
+            "params": {"NewSmartConnectEnable": value},
+            "body": "",
+            "need_auth": True
+        }
+
+        theResponse = self._set(theLog, theRequest, test)
+
+        return theResponse
+    #
```
[the remark this answers is not in the document]
Comment (roblandry) [reply pynetgear/__init__.py]: Fixed

File: pynetgear/helpers.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        "Content-Type":  "multipart/form-data"
+    }
+
+
+def is_valid_response(resp):
+    """Check if is valid."""
+    return (resp.status_code == 200 and
+            ("<ResponseCode>0000</ResponseCode>" in resp.text or
+             "<ResponseCode>000</ResponseCode>" in resp.text or
+             "<ResponseCode>2</ResponseCode>" in resp.text or  # Speed Test Result
```
[the remark this answers is not in the document]
Comment (roblandry) [reply pynetgear/helpers.py]: Fixed

File: pynetgear/helpers.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+             "<ResponseCode>001</ResponseCode>" in resp.text  # dns_masq/mac_address
```
[the remark this answers is not in the document]
Comment (roblandry) [reply pynetgear/helpers.py]: Fixed

**Comment by MatMaul** [comment]
Hi @roblandry, sorry for the inactivity on my side and thx for that.
I'll merge that soon, looks good to me. Can you just remove the Python 2.7 job on Travis and bump to 3.6 so the CI is happy ?

**Comment by starkillerOG** [comment]
Closing this in favor of the new PR: https://github.com/MatMaul/pynetgear/pull/120 which is heavily based on this PR and makes some aditional cleanups.

Thank you very very much for all the work @roblandry!

**Pull request closed without merging**